### PR TITLE
Add DynamoDB Streams integration for PartiQL query updates

### PR DESCRIPTION
**Pull request by whummer**
Add DynamoDB Streams integration for PartiQL query updates. Addresses #5445

* Add a handler/interceptor for `ExecuteStatement` API calls - these are [PartiQL queries](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ql-reference.statements.html) that potentially update the items in a table (via INSERT/UDPATE/DELETE queries). 
* The approach takes a snapshot of all table items prior to a PartiQL update operation, and compares them with the table items after the update. This will obviously have quite a performance impact for large tables, but seems to be the only viable option at this stage (as we're using the DynamoDB Local backend as a black box). Added a note to the code that this should be potentially be improved by hooking an update listener mechanism into the DynamoDB Local backend. However, given that this new code path is only executed if users are using DynamoDB Streams in combination with PartiQL update queries, the performance/simplicity tradeoff is probably acceptable for now (i.e., will only affect a very small subset of users). 
* Add a few smaller refactorings, start grouping util functions into logical groups/classes. This is only a first step - DynamoDB needs a _significant_ overhaul, which we'll tackle as part of the ASF migration (following soon).
* Migrate tests to `pytest`

Note: highlighted the 3 part with the main changes, for easier review.

### CODE REVIEW

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 1]
```diff
@@ -164,6 +262,16 @@ def forward_request(self, method, path, data, headers):
                 data["ReturnConsumedCapacity"] = "TOTAL"
                 return Request(data=json.dumps(data), method=method, headers=headers)
 
+        elif action == "ExecuteStatement":
```
Comment (whummer) [inline localstack/services/dynamodb/dynamodb_listener.py:265]: This part is new (rest of the PR mostly some smaller refactorings).

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 2]
```diff
@@ -379,6 +487,15 @@ def return_response(self, method, path, data, headers, response):
                 if stream_spec:
                     record["dynamodb"]["StreamViewType"] = stream_spec["StreamViewType"]
 
+        elif action == "ExecuteStatement":
```
Comment (whummer) [inline localstack/services/dynamodb/dynamodb_listener.py:490]: This part is new (rest of the PR mostly some smaller refactorings).

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 3]
```diff
@@ -942,51 +1048,62 @@ def has_event_sources_or_streams_enabled(table_name, cache=None):
...
-        for index in schema["Table"].get("GlobalSecondaryIndexes", []):
-            # TODO
-            # schemas.append(index['KeySchema'])
-            pass
-        for schema in schemas:
-            for key in schema:
-                key_name = key["AttributeName"]
-                search_key[key_name] = put_item["Item"][key_name]
-        if not search_key:
+def get_updated_records(table_name: str, existing_items: List) -> List:
```
Comment (whummer) [inline localstack/services/dynamodb/dynamodb_listener.py:1051]: This part is new (rest of the PR mostly some smaller refactorings).

**Comment by whummer** [comment]
Thanks for the pointer regarding running the records forwarding asynchronously in a background thread @steffyP ! 👍 PR updated, tests seem to be passing - ready for re-review.

**Review by steffyP — APPROVED** [review]
LGTM 👍 
And the refactoring part makes the code way more readable 😄

**Pull request merged**